cupboard_app/views.py

```diff
@@ -698,3 +712,82 @@ def list(self, request: Request) -> Response:
         all_measurements = get_all_measurements()
         serializer = MeasurementSerializer(all_measurements, many=True)
         return Response(serializer.data)
+
+
+@extend_schema(tags=['CustomIngredients'])
+class CustomIngredientsViewSet(viewsets.ViewSet):
+    MISSING_ING = 'Missing ingredient and type in message body.'
+
+    @extend_schema(
+        request=None,
+        responses={
+            201: CustomIngredientSerializer,
+            400: MessageSerializer,
+            401: auth_failed_response,
+        },
+        examples=[
+            OpenApiExample(
+                name='Custom Ingredient Created',
+                value={'name': 'Beef', 'type': 'Meat'},
+                status_codes=[201]
+            ),
+            OpenApiExample(
+                name='Required Value Missing',
+                value={'message': MISSING_ING},
+                status_codes=[400]
+            )
+        ]
+    )
+    def create(self, request: Request) -> Response:
+        """
+        Creates a custom ingredient for the user.
+        Returns the custom ingredient object.
+        """
+        # Extract username from the access token
+        username = get_auth_username_from_payload(request=request)
+        body = request.data
+        if (
+            username
+            and body.get('ingredient', None)
+            and body.get('type', None)
+        ):
+            custom_ingredient = create_custom_ingredient(
+                username=username,
+                name=body['ingredient'],
+                type=body['type']
+            )
+        else:
+            raise MissingInformation(self.MISSING_ING)
+
+        serializer = CustomIngredientSerializer(custom_ingredient)
+        return Response(serializer.data, status=201)
+
+    @extend_schema(
+        request=None,
+        responses={
+            200: CustomIngredientSerializer(many=True),
+            401: auth_failed_response,
+        },
+        examples=[
+            OpenApiExample(
+                name='Custom Ingredient Deleted',
+                value=CUSTOM_INGREDIENTS,
```

Comment:
Putting the custom ingredients as constant was good, one little caveat with the OpenAPI Swagger UI documentation is that when the response is set to `many=True` i.e. in line 768, it only wants an example of one of the items i.e. just one dictionary and not a list. Otherwise the example output becomes two nested arrays:
![image](https://github.com/user-attachments/assets/cfc5e13d-a9d5-433e-a9c6-288597219503)

So change the CUSTOM_INGREDIENTS to a single dictionary instead of array of dictionaries (i.e. do something like GROCERY_LIST.

Reply:
For some reason when I use ```value=[SINGLE_CUSTOM_INGREDIENT_DICT]``` the documentation shows a list inside a list like this:
```
[
  [
    {
      "user": "teacup",
      "name": "Beef",
      "type": "Meat"
    }
  ]
]
```
But if I do ```value=SINGLE_CUSTOM_INGREDIENT_DICT``` the documentation looks like this:
```
[
  {
    "user": "teacup",
    "name": "Beef",
    "type": "Meat"
  }
]
```

With ```SINGLE_CUSTOM_INGREDIENT_DICT``` declared as:
```
SINGLE_CUSTOM_INGREDIENT_DICT = {
        'user': 'teacup',
        'name': 'Beef',
        'type': 'Meat'
}
```
I assume we want option 2 here right?

Reply:
Oh yes, please go with option 2!